job-manager: allow jobtap plugins to reject jobs

grondo opened the pull request:
This PR changes the `job-manager.submit` protocol that job-ingest uses to submit batches of submitted jobs to the job manager. As described in #3472, job-ingest currently sends jobs in batches to the job manager in an all-or-nothing  approach, either the whole batch succeeds or it fails.

This PR extends that protocol to allow the job-manager to fail individual jobs. The response to `job-manager.submit` now may contain an optional `errors` key which is a list of `[jobid, errmsg]` tuples, each of which indicates that the given `jobid` failed. This allows one or more of a batch of jobs to be failed, each with a different error message.

Finally, a new jobtap hook `job.validate` is added which is called before a job is added to the job manager active jobs hash. A plugin can reject a job from this hook by returning `< 0`, with an optional `errmg` pushed onto the `FLUX_PLUGIN_ARG_OUT`. A convenience function, `flux_jobtap_reject_job ()` is added to allow a plugin to both push an error message onto the OUT args and return `-1` from the `job.validate` callback.

I tested this and it doesn't seem to horribly affect the ingest rate.

garlick commented on src/modules/job-manager/jobtap.c:
```diff
@@ -221,6 +221,38 @@ int jobtap_get_priority (struct jobtap *jobtap,
+    flux_plugin_arg_t *args;
+    const char *errmsg = NULL;
+
+    if (!jobtap->plugin)
+        return 0;
+    if (!(args = jobtap_args_create (jobtap, job)))
+        return -1;
```
Since the default plugin doesn't offer `job.validate`, should there maybe be test for  `flux_plugin_get_handler() == NULL` and an early return here to avoid a bit of extra work in `jobtap_args_create()`?

grondo replied:
Good point in general! This will be more tricky when we allow a list of plugins, but certainly helps now I think so I've added it.

garlick reviewed (approved):
Just one comment that you could ignore if it would provide insignificant improvement.  LGTM!

chu11 commented on src/modules/job-ingest/job-ingest.c:
```diff
@@ -117,6 +118,14 @@ struct batch {
 
+struct batch_response {
+    flux_future_t *f;
+    bool complete_failure;
```
nit: `job-info` uses the term `completed` for jobs that have finished running.  perhaps use a different flag name just to avoid potential confusion? `all_batch_failed` or similar?

chu11 commented on src/modules/job-manager/jobtap-internal.h:
```diff
@@ -39,6 +39,15 @@ int jobtap_get_priority (struct jobtap *jobtap,
                          struct job *job,
                          int64_t *pprio);
 
+
```
commit message typo: "An option error" -> 'An optional error"

chu11 commented on src/modules/job-manager/submit.c:
```diff
@@ -207,11 +261,15 @@ static void submit_cb (flux_t *h, flux_msg_handler_t *mh,
         goto error;
     }
-    if (flux_respond (h, msg, "{}") < 0)
+    if (flux_respond_pack (h, msg, "{s:o}", "errors", errors) < 0)
```
not a huge deal, but in `job-ingest` you have:

```
    /*  Empty payload indicates the whole batch was successful
     */
    if (o == NULL)
        return bresp;

    /*  O/w, there were one or more failures, capture these in the
     *   the response errors hash.
     */
```

it appears that you are always sending back an errors array, even if there are no errors?  Did you want to not send an array back when there are no errors?  Or perhaps just tweak the comment in `job-ingest`?

grondo replied:
The dummy job-manager module used in testing sends back an empty payload.

chu11 replied:
Ok, so I guess the only thing would be the comment "O/w there were one or more failures" in job-ingest.  B/c you may be sending back zero errors in the array?

grondo replied:
Got it. Good point, I didn't actually re-read the comment :facepalm:

chu11 commented on t/job-manager/plugins/test.c:
```diff
@@ -38,6 +48,7 @@ static int cb (flux_plugin_t *p,
                   "arg_pack: %s",
                   flux_plugin_arg_strerror (args));
 
+
```
nit: added a newline by accident :-)

chu11 reviewed (commented):
LGTM overall.  Mostly nits below, nothing serious.

grondo commented:
Here's an example plugin which will enforce a 16 job limit per user. This is just a simple demo to show how the interface can work
```C
#include <stdio.h>
#include <jansson.h>
#include <czmq.h>

#include <flux/core.h>
#include <flux/jobtap.h>

static int user_job_limit = 16;

struct user_info {
    uint32_t uid;
    int njobs;
};

static size_t uid_hasher (const void *key)
{
    const uint32_t *uid = key;
    return (size_t) *uid;
}

static void user_info_destructor (void **item)
{
    if (item) {
        free (*item);
        *item = NULL;
    }
}

static void user_hash_destroy (zhashx_t *uhash)
{
    if (uhash)
        zhashx_destroy (&uhash);
}

static zhashx_t *user_hash_create (void)
{
    zhashx_t *hash = zhashx_new ();
    if (!hash)
        return NULL;
    zhashx_set_key_hasher (hash, uid_hasher);
    zhashx_set_destructor (hash, user_info_destructor);
    return hash;
}

static int get_userid (flux_plugin_t *p,
                       flux_plugin_arg_t *args,
                       uint32_t *uidp)
{
    if (flux_plugin_arg_unpack (args,
                                FLUX_PLUGIN_ARG_IN,
                               "{s:i}",
                               "userid", uidp) < 0) {
        flux_t *h = flux_jobtap_get_flux (p);
        if (h)
            flux_log (h, LOG_ERR,
                      "flux_plugin_arg_unpack: %s",
                      flux_plugin_arg_strerror (args));
        return -1;
    }
    return 0;
}


static int validate (flux_plugin_t *p,
                     const char *topic,
                     flux_plugin_arg_t *args,
                     void *arg)
{
    zhashx_t *uhash = arg;
    struct user_info *ui;
    uint32_t userid;

    if (get_userid (p, args, &userid) < 0)
        return -1; /* Note: will reject job */

    if ((ui = zhashx_lookup (uhash, &userid))
        && ui->njobs == user_job_limit)
        return flux_jobtap_reject_job (p,
                                       args,
                                      "User reached job limit of %d jobs",
                                      user_job_limit);
    return 0;
}

static struct user_info * get_user_info (zhashx_t *uhash, uint32_t uid)
{
    struct user_info *ui;
    if (!(ui = zhashx_lookup (uhash, &uid)))  {
        if ((ui = calloc (1, sizeof (*ui))))
            ui->uid = uid;
        zhashx_insert (uhash, &ui->uid, ui);
    }
    return ui;
}

static int job_cb (flux_plugin_t *p,
                    const char *topic,
                    flux_plugin_arg_t *args,
                    void *arg)
{
    zhashx_t *uhash = arg;
    struct user_info *ui;
    uint32_t userid;
    flux_t *h = flux_jobtap_get_flux (p);

    if (get_userid (p, args, &userid) < 0) {
        flux_log (h, LOG_ERR, "Unable to get userid from args");
        return -1;
    }
    if (!(ui = get_user_info (uhash, userid))) {
        flux_log_error (h, "failed to create user info uid=%u", userid);
        return -1;
    }
    if (strcmp (topic, "job.new") == 0)
        ui->njobs++;
    else
        ui->njobs--;
    return 0;
}


int flux_plugin_init (flux_plugin_t *p)
{
    zhashx_t * uhash = user_hash_create ();
    if (!uhash) {
        flux_t *h = flux_jobtap_get_flux (p);
        if (h)
            flux_log (h, LOG_ERR, "plugin failed to create user hash");
        return -1;
    }
    flux_plugin_aux_set (p, NULL, uhash, (flux_free_f) user_hash_destroy);

    flux_plugin_set_name (p, "limiter");
    if (flux_plugin_add_handler (p, "job.validate", validate, uhash) < 0
        || flux_plugin_add_handler (p, "job.new", job_cb, uhash) < 0
        || flux_plugin_add_handler (p, "job.state.cleanup", job_cb, uhash))
        return -1;
    return 0;
}
```

e.g.:
```
 grondo@asp:~/git/f.git$ src/cmd/flux start 
ƒ(s=1,d=0,builddir) grondo@asp:~/git/f.git$ flux jobtap load test.so
Loaded:
limiter
Previously loaded:
builtin.priority.default
ƒ(s=1,d=0,builddir) grondo@asp:~/git/f.git$ flux mini submit --cc=1-18 hostname
ƒ87CXKRh
ƒ87CXKRi
ƒ87E1Ji3
ƒ87E1Ji4
ƒ87E1Ji5
ƒ87E1Ji6
ƒ87E1Ji7
ƒ87FVHzP
ƒ87FVHzQ
ƒ87FVHzR
ƒ87FVHzS
ƒ87FVHzT
ƒ87GyHGj
ƒ87GyHGk
ƒ87GyHGm
ƒ87GyHGn
cc=17: [Errno 22] User reached job limit of 16 jobs
cc=18: [Errno 22] User reached job limit of 16 jobs
```

grondo commented:
I force-pushed a new version with some fixes:

 * Misuse of zhashx @garlick helped me fix
 * added a `batch-count=N` test option to job-ingest to ensure we can test modes of batch failure in any test environment
 * fixed typos, improper variable name, and comment issue noted by @chu11

garlick commented:
Changes look good!

chu11 commented:
Changes LGTM too!

grondo commented:
Thanks! I'll set MWP.
